arch/lkl/kernel/threads.c

```diff
@@ -130,6 +133,19 @@ void free_thread_stack(struct task_struct *tsk)
 		test_tsk_thread_flag(tsk, TIF_SIGPENDING), ti, current->comm);
 
 	kill_thread(ti);
+	
+	/* free any linked list of pending signals */
+	/* note that we do not need to take the lock as the thread is dead and ought to be unreachable*/
+
+	spin_lock(&ti->signal_list_lock);
+	struct ksignal_list_node* signal_list = ti->signal_list_head;
+	while (signal_list != NULL) {
+		struct ksignal_list_node *next = signal_list->next;
+		kfree(signal_list);
+		signal_list = next;
+	}
+	spin_unlock(&ti->signal_list_lock);
```

Comment:
I wasn't sure whether **ti** could be referenced after the unlock so I was wondering if head and tail should be set to null.

Reply:
ti is kfree'd immediately after the list is destroyed. Some higher level logic must be keeping multiple threads away from the task structure.

Reply:
Thank you for the clarification.